### PR TITLE
ENH: Port bbregister T1-to-dwi registration from fmriprep

### PR DESCRIPTION
**Pull request by slimnsour**
Hello, this is to address #115 by adding bbregister to dmriprep. As you can see there is a little more work to be done because I wanted feedback on some things. Specifically:

1. Is there a best place I should put this workflow in? Ex. `dmriprep/workflows/dwi/base.py` or `dmriprep/workflows/dwi/registration.py`?
2. How should I import `init_bbreg_wf` in this case? I would have just imported `init_bbreg_wf` from fmriprep but fmriprep is not part of the packages installed in dmriprep. What is the suggested practice here?

Let me know of any feedback and I'll get to them.

### CODE REVIEW

File: dmriprep/cli/parser.py [hunk 1]
```diff
@@ -216,6 +216,23 @@ def _bids_filter(value):
...
+    )
+    g_conf.add_argument(
+        "--bold2t1w-dof",
+        action="store",
+        default=6,
+        choices=[6, 9, 12],
+        type=int,
+        help="Degrees of freedom when registering BOLD to T1w images. "
+        "6 degrees (rotation and translation) are used by default.",
+    )
```
Comment (oesteban) [inline dmriprep/cli/parser.py:235]: Let's not allow changing this. IMHO this should not exist in *fMRIPrep* either, we can call it legacy code that has not been cleaned up for historical reasons (and lack of bandwidth)
```suggestion
```

File: dmriprep/config/__init__.py [hunk 2]
```diff
@@ -436,6 +436,11 @@ class workflow(_Config):
     ignore = None
     """Ignore particular steps for *dMRIPrep*."""
     longitudinal = False
+    """Number of ICA components to be estimated by MELODIC
+    (positive = exact, negative = maximum)."""
+    bold2t1w_dof = None
+    """Degrees of freedom of the BOLD-to-T1w registration steps."""
+    bold2t1w_init = "register"
```
Comment (oesteban) [inline dmriprep/config/__init__.py:443]:
```suggestion
    bold2t1w_init = "register"
    """Whether to use standard coregistration ('register') or to initialize coregistration from the
    BOLD image-header ('header')."""
```
Although the PEP where attribute's docstrings were proposed was rejected, sphinx has maintained that style of documentation. The comment of the attribute goes AFTER the definition of the attribute.

Also, removed the dof option

File: dmriprep/workflows/base.py [hunk 3]
```diff
@@ -293,6 +293,30 @@ def init_single_subject_wf(subject_id):
...
     ])
 
+    if config.workflow.run_reconall:
+        from niworkflows.interfaces.nibabel import ApplyMask
+
+        # Mask the T1
+        t1w_brain = pe.Node(ApplyMask(), name='t1w_brain')
+
+        bbr_wf = init_bbreg_wf(use_bbr=True, bold2t1w_dof=config.workflow.bold2t1w_dof,
+                               bold2t1w_init=config.workflow.bold2t1w_init, omp_nthreads=config.nipype.omp_nthreads)
```
Comment (oesteban) [inline dmriprep/workflows/base.py:303]:
```suggestion
        bbr_wf = init_bbreg_wf(
            bold2t1w_init=config.workflow.bold2t1w_init,
            omp_nthreads=config.nipype.omp_nthreads,
            use_bbr=True,
        )
```

File: dmriprep/workflows/base.py [hunk 4]
```diff
@@ -293,6 +293,30 @@ def init_single_subject_wf(subject_id):
...
+
+        workflow.connect([
+            # T1 Mask
+            (anat_preproc_wf, t1w_brain, [('outputnode.t1w_preproc', 'in_file'),
+                                        ('outputnode.t1w_mask', 'in_mask')]),
```
Comment (oesteban) [inline dmriprep/workflows/base.py:308]:
```suggestion
            (anat_preproc_wf, t1w_brain, [('outputnode.t1w_preproc', 'in_file'),
                                          ('outputnode.t1w_mask', 'in_mask')]),
```

File: dmriprep/workflows/dwi/registration.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+
+from nipype.pipeline import engine as pe
+from nipype.interfaces import utility as niu, fsl, c3
+
+from ...interfaces import DerivativesDataSink
+
+DEFAULT_MEMORY_MIN_GB = config.DEFAULT_MEMORY_MIN_GB
+LOGGER = config.loggers.workflow
+
+def init_bbreg_wf(use_bbr, bold2t1w_dof, bold2t1w_init, omp_nthreads, name='bbreg_wf'):
```
Comment (oesteban) [inline dmriprep/workflows/dwi/registration.py:25]: I'm a bit unclear as to what changes w.r.t. *fMRIPrep* in this implementation

**Review by oesteban — COMMENTED** [review]
Looking great - I think we can import the `init_bbreg_wf` from *fMRIPrep* (I'll move it to smriprep or niworkflows, so that we do not depend on fmriprep). Please confirm.

File: dmriprep/cli/parser.py (hunk 1 above)
Comment (slimnsour) [reply dmriprep/cli/parser.py:235]: I was wondering about this, that clears it up.

File: dmriprep/config/__init__.py (hunk 2 above)
Comment (slimnsour) [reply dmriprep/config/__init__.py:443]: Thanks for catching this, I'll keep it in mind for the future.

File: dmriprep/workflows/dwi/registration.py (hunk 5 above)
Comment (slimnsour) [reply dmriprep/workflows/dwi/registration.py:25]: No changes from the fmriprep version here, we can just import it once it's in smriprep/niworkflows. I put it in here to get it working and I couldn't import it from fmriprep.

File: dmriprep/workflows/base.py [hunk 6]
```diff
@@ -293,6 +293,33 @@ def init_single_subject_wf(subject_id):
         ]),
     ])
 
+    if config.workflow.run_reconall:
+        from niworkflows.interfaces.nibabel import ApplyMask
+
+        # Mask the T1
```
Comment (oesteban) [inline dmriprep/workflows/base.py:299]:
```suggestion
        # Mask the T1w
```

A very-very stupid nitpick that helps me (and probably only me) sleep a bit better.

File: dmriprep/workflows/base.py [hunk 7]
```diff
@@ -293,6 +293,33 @@ def init_single_subject_wf(subject_id):
...
+        t1w_brain = pe.Node(ApplyMask(), name='t1w_brain')
+
+        bbr_wf = init_bbreg_wf(
+            bold2t1w_init=config.workflow.bold2t1w_init,
+            omp_nthreads=config.nipype.omp_nthreads,
+            use_bbr=True,
+        )
+
+        workflow.connect([
+            # T1 Mask
```
Comment (oesteban) [inline dmriprep/workflows/base.py:309]:
```suggestion
            # T1w Mask
```

File: dmriprep/workflows/base.py [hunk 8]
```diff
@@ -293,6 +293,33 @@ def init_single_subject_wf(subject_id):
...
+            (anat_preproc_wf, t1w_brain, [('outputnode.t1w_preproc', 'in_file'),
+                                          ('outputnode.t1w_mask', 'in_mask')]),
+            # BBregister
```
Comment (oesteban) [inline dmriprep/workflows/base.py:312]:
```suggestion
            # BBRegister
```

**Review by oesteban — COMMENTED** [review]
I'm going to add three stupid changes to your comments. Thanks for your understanding.

**Comment by oesteban** [comment]
@slimnsour please add your info to https://github.com/nipreps/dmriprep/blob/70028695186b5f4822a9328755c7f6791108c40a/.maint/developers.json#L17-L29

(it is ordered alphabetically by lastname)

**Comment by oesteban** [comment]
Shoot, I've overwritten your changes - assumed I was changing history (and I shouldn't have).

Can you add the lines here as a comment?

**Comment by slimnsour** [comment]
No worries, here it is.
 ```
{
    "affiliation": "The Centre for Addiction and Mental Health",
    "name": "Mansour, Salim",
    "orcid": "0000-0002-1092-1650"
},
```

**Comment by oesteban** [comment]
Works locally. I'm going to merge. Thanks for this great work @slimnsour and sorry for my slow response all the time.

**Pull request merged**